src/plugins/NEMD/MettDeamon.cpp

```diff
@@ -0,0 +1,2048 @@
+	}
+
+	double dVal = 0.0;
+	//keep storing values from the text file so long as data exists:
+	while (ifs.vxz >> dVal) {
+		_norm.vxz.push_back(dVal);
+	}
+	while (ifs.vy >> dVal) {
+		if(MD_LEFT_TO_RIGHT == _nMovingDirection)
+			_norm.vy.push_back( abs(dVal) );
+		else if (MD_RIGHT_TO_LEFT == _nMovingDirection)
+			_norm.vy.push_back( abs(dVal) * (-1.) );
+		else
+			Simulation::exit(-1);
+	}
+	// close files
+	ifs.vxz.close();
+	ifs.vy.close();
+}
+
+void MettDeamon::updateRandVecTrappedIns()
+{
+	create_rand_vec_ones(100, _feedrate.release_velo.normMB.a_neg, _feedrate.vec_rand_ins);
+//	cout << "_feedrate.vec_rand_ins: ";
+//	int nSum = 0;
+//	for(auto vi:_feedrate.vec_rand_ins) {
+//		cout << vi << ",";
+//		nSum += vi;
+//	}
+//	cout << "sum=" << nSum << endl;
+}
+
+//void MettDeamon::connectColleagues(const std::vector<MettDeamon*>& colleagues)
+//{
+//	if(_colleague.index < 1)
+//		return;
+//	if(_colleague.index-1 >= colleagues.size() )
+//		return;
+//	_colleague.ptr = colleagues.at(_colleague.index-1);
+//}
+
+//void MettDeamon::informColleagueAboutReleased()
+//{
+//	if(nullptr == _colleague.ptr)
+//		return;
+//	_colleague.ptr->IncrementInsertedMoleculesLocal();
+//}
+
+// mediator
+void MettDeamon::setParamsNormMB(const ParamsNormMB& params) {
+	_feedrate.release_velo.normMB.temperature = params.temperature;
+	_feedrate.release_velo.normMB.drift = params.drift;
+	_feedrate.release_velo.normMB.a_neg = params.a_neg;
+	_feedrate.release_velo.normMB.a_pos = params.a_pos;
+	_feedrate.release_velo.normMB.v_neg = params.v_neg;
+	_feedrate.release_velo.normMB.v_pos = params.v_pos;
+
+	DomainDecompBase& domainDecomp = global_simulation->domainDecomposition();
+	int nRank = domainDecomp.getRank();
+	if(0 == nRank || 1 == nRank) {
+		cout << "[" << nRank << "] MettDeamon: received new values: (T,D,a_neg,a_pos,v_neg,v_pos) = ";
+		cout << params.temperature << "," << params.drift << "," << params.a_neg << "," << params.a_pos << "," << params.v_neg << "," << params.v_pos << endl;
+	}
+
+	// update target feed rate
+	if(RVM_NORM_DISTR_GENERATOR == _feedrate.release_velo.method && FRM_CONSTANT == _nFeedRateMethod) {
+		double dts = 0.00182367;  // timestep of 2 fs
+		_feedrate.feed.target = abs(params.v_neg) * dts;
+		if(0 == nRank || 1 == nRank)
+			cout << "[" << nRank << "] New _feedrate.feed.target=" << _feedrate.feed.target << endl;
+	}
+
+	// update rand insertion vector to only release part of trapped particles
+	create_rand_vec_ones(100, _feedrate.release_velo.normMB.a_neg, _feedrate.vec_rand_ins);
+}
+
+// class Reservoir
+Reservoir::Reservoir(MettDeamon* parent) :
+	_parent(parent),
+	_moleculeDataReader(nullptr),
+	_binQueue(nullptr),
+	_numMoleculesRead(0),
+	_nMaxMoleculeID(0),
+	_nMoleculeFormat(ICRVQD),
+	_nReadMethod(RRM_UNKNOWN),
+	_dReadWidthY(0.0),
+	_dBinWidthInit(0.0),
+	_dBinWidth(0.0)
+{
+	// init filepath struct
+	_filepath.data = _filepath.header = "unknown";
+
+	// allocate BinQueue
+	_binQueue = new BinQueue();
```

Comment:
not released -> change to std::unique_ptr or make it stack variable!

Reply:
Here I got problems when using std::unique_ptr for the _binQueue. When the PluginFactory builds the MettDeamon plugin a Segmentation fault raised, or rather when the "default_delete" method for the _binQueue is called:

[simulator:13174] *** Process received signal ***
[simulator:13174] Signal: Segmentation fault (11)
[simulator:13174] Signal code: Address not mapped (1)
[simulator:13174] Failing at address: (nil)
[simulator:13174] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x10330)[0x7f6e586c4330]
[simulator:13174] [ 1] /home/mheinen/workspace/ls1mardyn_master/ls1-mardyn/src/MarDyn_43743e2.PAR_DEBUG_AVX(_ZN8BinQueueD1Ev+0x14)[0x69d032]
[simulator:13174] [ 2] /home/mheinen/workspace/ls1mardyn_master/ls1-mardyn/src/MarDyn_43743e2.PAR_DEBUG_AVX(_ZNKSt14default_deleteI8BinQueueEclEPS0_+0x22)[0x6a121a]
[simulator:13174] [ 3] /home/mheinen/workspace/ls1mardyn_master/ls1-mardyn/src/MarDyn_43743e2.PAR_DEBUG_AVX(_ZNSt10unique_ptrI8BinQueueSt14default_deleteIS0_EED1Ev+0x49)[0x69eec3]
...

Reply:
found the error:
```c++
~BinQueue() {
	_last->_next = nullptr;
	while (!isEmpty())
		deque();
}
```
in the destructor of BinQueue you are trying to set `_last->next` to `nullptr`. `_last` can, however, itself be a nullptr! So you get a segfault.

The more logical thing would be to change it to:
```c++
~BinQueue() {
	if (_last) {  // only do this if _last is not a nullptr!
		_last->_next = nullptr;
	}
	while (!isEmpty())
		deque();
}
```

But as far as I see it, your deque method was not used at all, so I implemented a more compact destructor, without the need of the dequeue method:

```c++
~BinQueue() {
	if(_last){
		// break connection of last element to first element (established via connectTailToHead())
		_last->_next = nullptr;
	}
	auto ptr = _first;
	while (ptr) {
		// as long as ptr isn't nullptr we continue to delete the next element.
		auto ptr_next = ptr->_next;
		delete ptr;
		ptr = ptr_next;
	}
}
```